privacy/bolt_on/losses_test.py

```diff
@@ -242,11 +236,162 @@ def test_prints(self, init_args, fn, args, print_res):
       loss = StrongConvexBinaryCrossentropy(*init_args)
       if fn is not None:
         getattr(loss, fn, lambda *arguments: print('error'))(*args)
-    self.assertRegexMatch(err.getvalue().strip(), [print_res])
+    self.assertRegexMatch(out.getvalue().strip(), [print_res])
+
+class CategoricalCrossesntropyTests(keras_parameterized.TestCase):
+  """tests for CategoricalCrossesntropy StrongConvex loss."""
+
+  @parameterized.named_parameters([
+      {'testcase_name': 'normal',
+       'reg_lambda': 1,
+       'C': 1,
+       'radius_constant': 1
+      },  # pylint: disable=invalid-name
+  ])
+  def test_init_params(self, reg_lambda, C, radius_constant):
+    """Test initialization for given arguments.
+    Args:
+      reg_lambda: initialization value for reg_lambda arg
+      C: initialization value for C arg
+      radius_constant: initialization value for radius_constant arg
+    """
+    # test valid domains for each variable
+    loss = StrongConvexCategoricalCrossentropy(reg_lambda, C, radius_constant)
+    self.assertIsInstance(loss, StrongConvexCategoricalCrossentropy)
 
+  @parameterized.named_parameters([
+      {'testcase_name': 'negative c',
+       'reg_lambda': 1,
+       'C': -1,
+       'radius_constant': 1
+      },
+      {'testcase_name': 'negative radius',
+       'reg_lambda': 1,
+       'C': 1,
+       'radius_constant': -1
+      },
+      {'testcase_name': 'negative lambda',
+       'reg_lambda': -1,
+       'C': 1,
+       'radius_constant': 1
+      },  # pylint: disable=invalid-name
+  ])
+  def test_bad_init_params(self, reg_lambda, C, radius_constant):
+    """Test invalid domain for given params. Should return ValueError.
+    Args:
+      reg_lambda: initialization value for reg_lambda arg
+      C: initialization value for C arg
+      radius_constant: initialization value for radius_constant arg
+    """
+    # test valid domains for each variable
+    with self.assertRaises(ValueError):
+      StrongConvexCategoricalCrossentropy(reg_lambda, C, radius_constant)
+
+  @test_util.run_all_in_graph_and_eager_modes
+  @parameterized.named_parameters([
+      # [] for compatibility with tensorflow loss calculation
+      {'testcase_name': 'both positive',
+       'logits': [10000, 0],
```

Comment:
Is this the right test case for CategoricalCrossentropy? 

According to https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/losses/CategoricalCrossentropy, the input for the loss function (logits and labels) should be two dimensional (batch_size, num_classes). Secondly, the input should be valid probability distributions. Here is the example given by the doc:
```python
cce = tf.keras.losses.CategoricalCrossentropy()
loss = cce(
  [[1., 0., 0.], [0., 1., 0.], [0., 0., 1.]],
  [[.9, .05, .05], [.05, .89, .06], [.05, .01, .94]])
print('Loss: ', loss.numpy())  # Loss: 0.0945
```

Reply:
@cchoquette what do you think? I was trying to replicate the test cases from binary cross entropy. The logits being proper probabilities (that sum up to 1) do not affect the matematical tests.

Reply:
This is fine, just use from_logits=True.
I would then also add some test cases with from_logits=False which should have their inputs be valid pdfs.
I see you did use from_logits=True as the default and I think it's better, especially in the high-level model API (it could be more numerically stable and quicker as TF does optimizations in the backend when combining the softmax and crossentropy losses together).